File: ios/Adjutant/Sources/Features/Chat/MessageGrouping.swift
import Foundation
import AdjutantKit

/// Same-sender run grouping + auto-scroll policy for the DM timeline
/// (adj-164.3.3). Parity with the web `messageGrouping.ts` + CommandChat
/// `followOutput` so iOS and web render bubbles identically.
///
/// Consecutive messages from the same sender collapse into a visual "run":
/// only the FIRST message in a run shows the sender callsign header, only the
/// LAST shows the timestamp / delivery status. This reads as one continuous
/// transmission instead of a stack of identically-chromed boxes.
///
/// Pure and O(n): a single pass comparing each message's sender key to its
/// neighbours. Kept out of the view so it is unit-testable in isolation.
enum MessageGrouping {

    /// Per-message grouping flags.
    struct GroupFlags: Equatable {
        /// First message in a same-sender run → render the sender callsign.
        let isFirstInGroup: Bool
        /// Last message in a same-sender run → render the timestamp / status.
        let isLastInGroup: Bool
    }

    /// The identity a run is keyed on. The user is always one sender; each
    /// distinct agent is its own sender (two agents never share a run). System
    /// and announcement rows get a unique key so they never group.
    private static func senderKey(_ msg: PersistentMessage) -> String {
        switch msg.role {
        case .user:
            return "user"
        case .system, .announcement:
            // Never group system/announcement rows — make every key unique.
            return "__system__\(msg.id)"
        case .agent:
            return "agent:\(msg.agentId)"
        }
    }

    /// Compute per-message grouping flags for a chronologically-ordered list
    /// (oldest → newest).
    ///
    /// - Parameter messages: messages in display order.
    /// - Returns: a dictionary keyed by message id with each message's flags.
    static func computeGroups(for messages: [PersistentMessage]) -> [String: GroupFlags] {
        var flags: [String: GroupFlags] = [:]
        flags.reserveCapacity(messages.count)

        for i in messages.indices {
            let current = messages[i]
            let key = senderKey(current)

            let prevKey = i > 0 ? senderKey(messages[i - 1]) : nil
            let nextKey = i < messages.count - 1 ? senderKey(messages[i + 1]) : nil

            let isFirst = prevKey == nil || prevKey != key
            let isLast = nextKey == nil || nextKey != key

            flags[current.id] = GroupFlags(isFirstInGroup: isFirst, isLastInGroup: isLast)
        }

        return flags
    }

    /// Auto-scroll policy: pin to the newest message only when the user is
    /// already at the bottom. When they've scrolled up to read history, return
    /// false so the timeline does not jump out from under them. Mirrors the web
    /// `followOutput(isAtBottom)` predicate.
    static func shouldAutoScroll(isAtBottom: Bool) -> Bool {
        isAtBottom
    }
}

File: frontend/src/components/chat/messageGrouping.ts
/**
 * messageGrouping — same-sender run grouping for the chat timeline (adj-164.2.2).
 *
 * Consecutive messages from the same sender are collapsed into a visual "run":
 * only the FIRST message in a run shows the sender callsign, only the LAST
 * shows the timestamp/delivery status, and the bubbles in between sit closer
 * together. This reads like one continuous terminal transmission rather than a
 * stack of identically-chromed boxes.
 *
 * The function is pure and O(n): a single pass comparing each message's sender
 * key to its neighbours. Extracted from CommandChat so it is unit-testable in
 * isolation and so the virtualized render path can call it once per message
 * list (memoized) without dragging in component state.
 */

import type { DisplayMessage } from "../../hooks/useChatMessages";

export interface MessageGroupFlags {
  /** First message in a same-sender run → render the sender callsign header. */
  isFirstInGroup: boolean;
  /** Last message in a same-sender run → render the timestamp / delivery status. */
  isLastInGroup: boolean;
}

/**
 * The identity a run is keyed on. The user is always one sender; each distinct
 * agent is its own sender (two different agents never share a run). System and
 * announcement rows are deliberately given unique keys so they never group.
 */
function senderKey(msg: DisplayMessage): string {
  if (msg.role === "user") return "user";
  if (msg.role === "system" || msg.role === "announcement") {
    // Never group system/announcement rows — make every key unique.
    return `__system__${msg.id}`;
  }
  return `agent:${msg.agentId}`;
}

/**
 * Compute per-message grouping flags for a chronologically-ordered list.
 *
 * @param messages messages in display order (oldest → newest)
 * @returns a Map keyed by message id with each message's group flags
 */
export function computeMessageGroups(
  messages: DisplayMessage[],
): Map<string, MessageGroupFlags> {
  const flags = new Map<string, MessageGroupFlags>();

  for (let i = 0; i < messages.length; i++) {
    const current = messages[i];
    if (!current) continue;
    const key = senderKey(current);

    const prev = messages[i - 1];
    const next = messages[i + 1];

    const isFirstInGroup = !prev || senderKey(prev) !== key;
    const isLastInGroup = !next || senderKey(next) !== key;

    flags.set(current.id, { isFirstInGroup, isLastInGroup });
  }

  return flags;
}
